tests/koops_stacktrace.c

```diff
@@ -1,16 +1,794 @@
+#include "koops/stacktrace.h"
+#include "location.h"
+#include "utils.h"
+#include "thread.h"
+#include "stacktrace.h"
+#include <stdio.h>
 #include <glib.h>
 
+static void
+test_koops_stacktrace_parse_modules(void)
+{
+    struct test_data
+    {
+        bool success;
+        char *input;
+        char **expected_modules;
+    };
+
+    struct test_data t[] = {
+    };
+
+    for (int i = 0; i < sizeof (t) / sizeof (*t); i++)
+    {
+        char *old_input = t[i].input;
+        char **modules = sr_koops_stacktrace_parse_modules((const char **)&t[i].input);
+        g_assert_true(!!modules == t[i].success);
+
+        if (modules)
+        {
+            int offset = 0;
+            while (modules[offset])
+            {
+                g_assert_nonnull(t[i].expected_modules[offset]);
+                g_assert_cmpstr(modules[offset], ==, t[i].expected_modules[offset]);
+                ++offset;
+            }
+            g_assert_null(t[i].expected_modules[offset]);
+
+            offset = 0;
+            while (modules[offset])
+            {
+                free(modules[offset]);
+                ++offset;
+            }
+
+            free(modules);
+        }
+        else
+        {
+            /* Check that the pointer is not moved. */
+            g_assert_true(old_input == t[i].input);
+        }
+    }
+}
+
+static void
+test_koops_stacktrace_parse(void)
+{
+    struct test_data
+    {
+        const char *path;
+        const char *first_function_name;
+        const char *last_function_name;
+        int frame_count;
+        int taint_count;
+        int module_count;
+        char *first_frame_stack;
+        char *last_frame_stack;
+    };
+
+    struct test_data t[] = {
+        {"kerneloopses/gitlog-01", "unwind_backtrace", "kthread", 15, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-02", "unwind_backtrace", "cpu_idle", 27, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-03", "lockdep_rcu_suspicious", "system_call_fastpath", 19, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-04", "lockdep_rcu_suspicious", "addrconf_forward_change", 7, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-05", "mark_held_locks", "sysenter_do_call", 26, 1, 0, NULL, NULL},
+        {"kerneloopses/gitlog-06", "lock_acquire", "kernel_thread_helper", 23, 1, 0, NULL, NULL},
+        {"kerneloopses/gitlog-07", "__slab_alloc.isra.50.constprop.56", "sysenter_do_call", 14, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-08", "unwind_backtrace", "sys_write", 13, 1, 0, NULL, NULL},
+        {"kerneloopses/gitlog-09", "unwind_backtrace", "cpu_idle", 11, 1, 0, NULL, NULL},
+        {"kerneloopses/gitlog-10", NULL, NULL, 0, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-11", "lockdep_rcu_suspicious", "out_of_memory", 12, 0, 0, NULL, NULL},
+        {"kerneloopses/gitlog-12", "unwind_backtrace", "regulator_get_voltage", 6, 1, 0, NULL, NULL},
+
+        /*
+        {"kerneloopses/gitlog-13", "", "", 0, 0, 0, NULL, NULL},
```

Comment:
Commented code can be removed.

Reply:
Hm, some test data seems to be ready in `tests/kerneloopses`. I wonder if it's worth the effort to fill in the structs properly and use it in the test instead of just dropping all this.

Reply:
Well, let's try and see provided it's not too much hassle. There can never be too many tests.

Reply:
I tried first few files but can't get it to work. Moreover I feel like I'm reverse-engineering the test data to design a "test" that succeeds, more than anything else. I suppose it can be put to good use but I'd rather leave it to somebody who knows what they're doing.